Make Screen Curtain profile independent

SaschaCowley opened the pull request:
### Link to issue number:

Fixes #10476
Partially addresses #16272

### Summary of the issue:

Currently, the Screen Curtain is configuration profile dependent. This can lead to situations in which it is unexpectedly deactivated by a config profile change.

### Description of user facing changes:

1. The Screen Curtain is no-longer config profile dependent.
2. Screen Curtain settings have been moved to anew "Privacy and Security" settings category.

### Description of developer facing changes:

The `visionEnhancementProviders.screenCurtain` module has bee removed.

### Description of development approach:

* Added a new `screenCurtain` subpackage, and migrated the existing `visionEnhancementProviders.screenCurtain` implementation.
  * The global object pattern was used, as it is a common pattern throughout the rest of NVDA's code.
  * The core Screen Curtain code is in `screenCurtain._screenCurtain`. The public API is accessible from `screenCurtan`, including the global `screenCurtain.screenCurtain` object.
* Updated `core.main` and `core.resetConfiguration` to initializeand terminate screen curtain.
* Updated the screen curtain toggle script to use the new implementation.
* Updated the OCR script and screen curtain block action.
* Re-implemented the screen curtain settings as regular settings in a new "Privacy and Security" panl in `gui.settingsDialogs`.
  * I decided to call this category "Privacy and Security" rather than "Security and Privacy" (as suggested in #16272), as we already have a "Speech" category, but no categories beginnin wit 'P'.
  * Also added an unassigned command to jump directly to the privacy and security settings.
* Completely removed the `visionEnhancementProviders.screenCurtain` module, and updated the change log to reflect that previously deprecated symbols that had redirects in this module have been removed.
* Bumped the config schema version, and added a profile upgrade step that moves the settings out of `vision` and into the root section. As the config keys are otherwise identical, this is all that is necessary.
* Updated changed strings to use `pgettext` (category "screenCurtain")

### Testing strategy:

Running from source:

* Enabled tempory screen curtain with the gesture, with and without the warning enabled. Disabled.
* Enabled screen curtain with the gesture, with and without the dialog enabled. Disabled.
* Checked and uncheckd the screen curtain option in Privacy and Security settings, with and without the warning enabled. Also checked that unchecking the option to show a warning was reflected in the settings dialog when the warning dialog was shown from NVDA settings.
* Enabled and disabled Screen Curtain with sounds turned on and off.
* Created a configuration profile. Enabled screen curtain. Saved settings. Activated the profile. Deactivated screen curtain. Saved settings. Switched back to normal config.
* Reset to saved configuration after enabling and disabling screen curtain, and reset configuration to factory defaults.
* Ran from source on master, and changed screen curtain settings. Ran from source on this branch and ensured the old settings were reflected. Saved the config to disk and ensured that the settings were correctly persisted to `nvda.ini`.

### Known issues with pull request:

None

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

SaschaCowley committed:
Add new screenCurtain module and start initial development

SaschaCowley committed:
Einitializer and deleter

SaschaCowley committed:
Initialize and terminate wit NVDA

SaschaCowley committed:
Update script

SaschaCowley committed:
Improve OCR and screen curtain scripts

SaschaCowley committed:
Remove winVersion.isFullScreenMagnificationAvailable

SaschaCowley committed:
Fix block action

SaschaCowley committed:
Settings dialog

SaschaCowley committed:
Tidy API

SaschaCowley committed:
Fix settings

SaschaCowley committed:
Actually make profile independent, and rename settings category

SaschaCowley committed:
Update documentation

SaschaCowley committed:
Don't treat visionEnhancementProviders.screenCurtain as a vision enhancement provider

SaschaCowley committed:
Add to changes

SaschaCowley committed:
Restart screen curtain when resetting config

SaschaCowley committed:
Update references to screen curtain

cary-rowen commented:
#10476 is too long for me to finish reading, but I just want to confirm one thing: will this change cause the screen curtain to not follow the process-name-based profile? If so, that would be terrible. I have a very important use case: I created separate profiles for my instant messaging applications, such as "WeChat," because I don't want these applications to be displayed in specific scenarios, while for other applications I don't think it matters much, so I don't want to enable the screen curtain.

SaschaCowley committed:
Misc fixups

SaschaCowley committed:
Add some docstrings

SaschaCowley commented on source/visionEnhancementProviders/screenCurtain.py:
Should we just remove this module all together?

SaschaCowley commented on source/screenCurtain.py:
```diff
@@ -0,0 +1,234 @@
+
+# Translators: Description for a Screen Curtain setting that shows a warning when enabling
+# the screen curtain.
+warnOnLoadCheckBoxText = _("Always &show a warning when enabling Screen Curtain")
```
Can we mark this as an internal constant?

SaschaCowley committed:
Update changes with PR number

SaschaCowley commented on source/screenCurtain.py:
```diff
@@ -0,0 +1,234 @@
+
+# homogeneous matrix for a 4-space transformation (red, green, blue, opacity).
+# https://docs.microsoft.com/en-gb/windows/win32/gdiplus/-gdiplus-using-a-color-matrix-to-transform-a-single-color-use
+TRANSFORM_BLACK = magnification.MAGCOLOREFFECT()  # empty transformation
```
Is there a reason to have this be part of the public API?

seanbudd replied:
probably worth removing

seanbudd commented on source/core.py:
```diff
@@ -811,6 +816,10 @@ def main():
+	import screenCurtain
+
+	log.debug("Initializing Screen Curtain")
+	screenCurtain.initialize()
```
```suggestion

	import screenCurtain

	log.debug("Initializing Screen Curtain")
	screenCurtain.initialize()

```

SaschaCowley committed:
Add unassigned gesture to jump directly to privacy and security settings

SaschaCowley committed:
Add profile update steps

SaschaCowley committed:
mend

SaschaCowley committed:
Update changes

CyrilleB79 commented:
> > If so, that would be terrible. I have a very important use case: I created separate profiles for my instant messaging applications, such as "WeChat," because I don't want these applications to be displayed in specific scenarios, while for other applications I don't think it matters much, so I don't want to enable the screen curtain.
> 
> Unfortunately, as things stand there is no solution that will please everyone. We have made the decision to make the screen curtain profile independent as we feel that this provides the best experience for users in terms of privacy. Perhaps as NVDA's configuration profile support improves this will be a use-case that can be added back.

It's not even a matter of pleasing everyone, or satisfying different use cases.
It's just that a profile dependant, so an app-dependant screen curtain gives a false sense of privacy but cannot guarantee it actually.

The privacy issue is as follows:
Let's imagine you have the base profile without the screen curtain off and an IM app profile, e.g. WeChat profile, with screen curtain on. If you are in WeChat and WeChat window is big enough and you open another app with a smaller window, parts of the bigger window will be visible in the background since the smaller window will not cover the whole IM window in the background. I guess you can reproduce by maximizing WeChat window and opening the "Run" window with `windows+R`.

seanbudd commented on source/globalCommands.py:
```diff
@@ -4801,19 +4807,20 @@ def script_speech_cycleUnicodeNormalization(self, gesture: inputCore.InputGestur
+		if screenCurtain is None:
+			# Screen curtain has not been initialized.
+			# Translators: Reported when the screen curtain is not available.
+			ui.message(_("Screen curtain not available"), speechPriority=speech.priorities.Spri.NOW)
```
```suggestion
			ui.message(pgettext("screenCurtain", "Screen curtain not available"), speechPriority=speech.priorities.Spri.NOW)
```

seanbudd commented on source/gui/settingsDialogs.py:
```diff
@@ -5585,6 +5586,114 @@ def onSave(self):
+			wx.CheckBox(
+				screenCurtainBox,
+				#  Translators: option to enable screen curtain in the privacy and security settings panel
+				label=_("Make screen black (immediate effect)"),
```
```suggestion
				label=pgettext("screenCurtain", "Make screen black (immediate effect)"),
```

seanbudd commented on source/screenCurtain.py:
```diff
@@ -0,0 +1,234 @@
+		self,
+		screenCurtainSettingsStorage: _ScreenCurtainSettings,
+		parent: wx.Window,
+		title: str = _("Warning"),
```
add translator comment please

seanbudd commented on source/screenCurtain.py:
```diff
@@ -0,0 +1,234 @@
+	global screenCurtain
+	if screenCurtain is not None:
+		screenCurtain.disable(persist=False)
+		screenCurtain = None
```
could we consider making this a module and just keeping this part in the __init__.py?

CyrilleB79 commented:
@SaschaCowley wrote:
> The `visionEnhancementProviders.screenCurtain` module is essentially just a stub now. If we are happy breaking the few redirects in that module, we can get rid of it, and the code excluding it when loading VEPs from the file system.

In the few add-ons that seem to use screen curtain API, most seem to be just testing if screen curtain is active or not to perform OCR, recognition, etc. with the following code (or similar):

```
import vision
from visionEnhancementProviders.screenCurtain import ScreenCurtainProvider
screenCurtainId = ScreenCurtainProvider.getSettings().getId()
screenCurtainProviderInfo = vision.handler.getProviderInfo(screenCurtainId)
isScreenCurtainActive = bool(vision.handler.getProviderInstance(screenCurtainProviderInfo))
```

 Thus, I do not expect them to have difficulty to adapt to the new code if necessary.

Since 2026.1 will already be an important API-breaking release, I'd be in favor to take the opportunity to simplify NVDA's code here and remove the stub, rather than keeping this deprecated stub for only few add-ons.

SaschaCowley committed:
Apply suggestions from code review

Co-authored-by: Sean Budd <sean@nvaccess.org>

SaschaCowley committed:
Refactor screenCurtain to be a package with an internal _screenCurtain module containing all internal API

SaschaCowley committed:
Make duplicated message an internal constant

SaschaCowley committed:
Rename to be constant

SaschaCowley committed:
Add more thorough type hints

SaschaCowley committed:
Fix toggle command

SaschaCowley committed:
Make imports global

SaschaCowley committed:
Remove visionEnhancementProviders.screenCurtain altogether

SaschaCowley committed:
Use pgettext

CyrilleB79 commented on source/globalCommands.py:
```diff
@@ -4852,28 +4862,21 @@ def script_toggleScreenCurtain(self, gesture):
 		):
 			# Translators: Reported when the screen curtain is disabled.
-			message = _("Screen curtain disabled")
+			message = pgettext("screenCurtain", "Screen curtain disabled")
```
Her and elsewhere in this PR, you have provided context to translatable strings, even where the strings had not been modified. The consequence is that all these strings will have to be translated again by translators.

The reasons why we provide a context info for translatable strings (usage of `pgettext` instead of `_`) are:
1. Allow to provide 2 different translations, depending on the context, for a same string in English. E.g. "off" or "None" may be translated differently depending on the situation in which they are used.
2. Allows to provide a context information for translators (even if the translators comments can also provide useful information)

For unmodified string curtain strings that are already translated for long, reason 2. does not apply, and there is just the drawback for translators to have to translate them again.

So unless there are real situation where we want to distinguish 2 identical strings to have them translated differently for different usages / contexts, I'd recommend not to convert `_` to `pgettext`, neither in this PR nor in other ones.

Though if a string is modified for whatever reason, it may be the opportunity to add a context element.

seanbudd commented on source/screenCurtain/__init__.py:
```diff
@@ -0,0 +1,33 @@
+# A part of NonVisual Desktop Access (NVDA)
```
can this get the latest version of copyrightHeaders.md

seanbudd commented on source/screenCurtain/_screenCurtain.py:
```diff
@@ -0,0 +1,221 @@
+# A part of NonVisual Desktop Access (NVDA)
```
can this get the latest version of copyrightHeaders.md?

seanbudd commented on user_docs/en/changes.md:
```diff
@@ -50,6 +50,8 @@ We recommend using Windows 11, or if that is not possible, the latest Windows 10
   * Updated BrlAPI for BRLTTY to version 0.8.7, and its corresponding python module to a Python 3.13 compatible build. (#18657, @LeonarddeR)
 * In browse mode in web browsers, NVDA no longer sometimes treats controls with 0 visual width or height as invisible. This technique is sometimes used to make content accessible to screen readers without it being visible visually. Such controls will now be accessible in browse mode where they weren't before. (#13897, @jcsteh)
+* The state of the Screen Curtain is no-longer dependent on the configuration profile in use. (#10476)
+* The screen Curtain settings have been moved to the new privacy and security category of NVDA's settings. (#19177)
```
```suggestion
* Screen Curtain settings have been moved to the new Privacy and Security category of NVDA's settings. (#19177)
```

seanbudd commented on user_docs/en/userGuide.md:
```diff
@@ -1424,7 +1424,7 @@ As a blind or vision impaired user, it is often not possible or necessary to see
 For this situation, NVDA contains a feature called "Screen Curtain" which can be enabled to make the screen black.
 
-You can enable the Screen Curtain in the [vision category](#VisionSettings) of the [NVDA Settings](#NVDASettings) dialog.
+You can enable the Screen Curtain in the [privacy and security category](#PrivacyAndSecuritySettings) of the [NVDA Settings](#NVDASettings) dialog.
```
```suggestion
You can enable the Screen Curtain in the [Privacy and Security category](#PrivacyAndSecuritySettings) of the [NVDA Settings](#NVDASettings) dialog.
```

seanbudd commented on user_docs/en/userGuide.md:
```diff
@@ -1434,11 +1434,7 @@ You can enable the Screen Curtain in the [vision category](#VisionSettings) of t
-Due to a change in the Windows Magnification API, Screen Curtain had to be updated to support the newest versions of Windows.
-Use NVDA 2021.2 to activate Screen Curtain with Windows 10 21H2 (10.0.19044) or later.
-For security purposes, when using a new version of Windows, get visual confirmation that the Screen Curtain makes the screen entirely black.
+When the Screen Curtain is active, some tasks directly based on what appears on the screen, such as performing [OCR](#Win10Ocr) or taking a screenshot, cannot be achieved.
```
```suggestion
When the Screen Curtain is active, some tasks directly based on what appears on the screen, such as performing [OCR](#Win10Ocr), [image descriptions]() or taking a screenshot, cannot be achieved.
```

SaschaCowley committed:
Update copyright headers

SaschaCowley committed:
Apply suggestions from code review

Co-authored-by: Sean Budd <sean@nvaccess.org>

SaschaCowley committed:
Update user guide

SaschaCowley committed:
Switch back to plain gettext for unchanged strings

CyrilleB79 reviewed (approved):
Nice! All good for me.

seanbudd reviewed (approved):
Thanks @SaschaCowley - happy with this too. I have some suggestions, but I'm not sure how appropriate they are

Qchristensen committed:
Update user_docs/en/userGuide.md

I think because we have called the feature "Screen Curtain" (and it's capitalised like feature names we've used) - but the words themselves are common and could simply be written in a sentence which would make sense even if we didn't have such a feature (and in that case you would use "the").  If we change this, we should probably change other references to match.

Co-authored-by: Sean Budd <sean@nvaccess.org>

Qchristensen commented on user_docs/en/userGuide.md:
```diff
@@ -2618,6 +2614,36 @@ Setting the value too high may cause the battery of the sound output device to d
+##### Make screen black (immediate effect) {#ScreenCurtainEnable}
+
+This option enables and disables the [Screen Curtain](#VisionScreenCurtain).
+Note that the Screen Curtain is activated or deactivated as soon as the checkbox is checked or unchecked.
```
```suggestion
Note that Screen Curtain is activated or deactivated as soon as the checkbox is checked or unchecked.
```

Qchristensen commented on user_docs/en/userGuide.md:
```diff
@@ -2618,6 +2614,36 @@ Setting the value too high may cause the battery of the sound output device to d
+
+This option controls whether a warning dialog is displayed when enabling the [Screen Curtain](#VisionScreenCurtain).
+
+If you use the Screen Curtain regularly, you may find it more convenient to disable this option.
```
```suggestion
If you use Screen Curtain regularly, you may find it more convenient to disable this option.
```

Qchristensen commented on user_docs/en/userGuide.md:
```diff
@@ -2618,6 +2614,36 @@ Setting the value too high may cause the battery of the sound output device to d
+
+By default, sounds are played when the Screen Curtain is toggled.
+If you do not wish to hear these sounds, you can uncheck this checkbox.
+Note that you will still be alerted in speech and/or braille when the Screen Curtain is toggled with an input gesture.
```
```suggestion
Note that you will still be alerted in speech and/or braille when Screen Curtain is toggled with an input gesture.
```

Qchristensen reviewed (approved):
Reads well. Following Sean's suggestion about dropping "the" from "the Screen Curtain", I've changed other references in the user guide (I note there other uses of "the Screen Curtain" in the code comments but haven't changed those)

SaschaCowley commented on source/screenCurtain/_screenCurtain.py:
```diff
@@ -0,0 +1,220 @@
+# the screen curtain.
+WARN_ON_LOAD_CHECKBOX_TEXT = pgettext("screenCurtain", "Always &show a warning when enabling Screen Curtain")
+
+UNAVAILABLE_WHEN_RECOGNISING_CONTENT_MESSAGE = pgettext(
```
I think moving it to blockAction would make implementing #19164 rather difficult.

SaschaCowley committed:
Apply suggestions from code review

Co-authored-by: Quentin Christensen <quentin@nvaccess.org>

SaschaCowley committed:
Update user_docs/en/userGuide.md
